source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -238,7 +299,7 @@ Network::IoResult SslSocket::doWrite(Buffer::Instance& write_buffer, bool end_st
 void SslSocket::onConnected() { ASSERT(!handshake_complete_); }
 
 void SslSocket::shutdownSsl() {
-  ASSERT(handshake_complete_);
+  ASSERT(handshake_complete_ || async_handshake_in_progress_);
```

Comment:
If there is a async handshake in progress, then we should notify `PrivateKeyConnection` or `PrivateKeyMethodProvider` that this connection is going away and that we're no longer interested in the result, so that more complex providers can cancel the costly computation.

Probably in `SslSocket::closeSocket` and not here, though.

Reply:
Yes. The original idea was to tie the object lifecycles so that the `PrivateKeyConnection` destructor can cancel the operation. It might be better to do it explicitly though to prevent any races.

Reply:
@PiotrSikora: I added the unregistering now to `SslSocket::closeSocket()`. However, I'm not sure if the state machine will always call it; it seems that `SslSocket::doWrite()` can directly call `shutdownSsl()` too. Would it be safer that unregistering would be part of `SslSocket` destructor, and we would just ignore any async callbacks after `closeSocket()` and before the object destruction?

Reply:
I believe it will always call `closeSocket()`.

Reply:
Ok, this is good.